### PR TITLE
feat: reset marks user anonymous

### PR DESCRIPTION
**Pull request by pauldambra**
## Changes

In our docs we say "If a user is logged out, you most likely want to call reset to unset any distinct_ids."

There are several instances of this confusing users see https://github.com/PostHog/posthog-js/issues/512

users expect 

`identify person 1 -> event A -> reset -> event B -> identify person 2`

to result in `event B` being associated with `person 2`

but that doesn't happen because we are guarding against `identify person 1 -> event A -> event B -> identify person 2`

(Note no reset in this chain)

----

To get the expected behaviour you have to call `reset(true)`

----

Because we are already using state to track a proxy for whether a user is anonymous we can use a little more state to directly track whether a user is anonymous.

This allows `identify -> event -> identify -> event` and `identify -> event -> reset -> identify -> event` to behave as expected

## Checklist
- [x] Tests for new code (see [advice on the tests we use](https://github.com/PostHog/posthog-js#tiers-of-testing))
- [x] Accounted for the impact of any changes across different browsers

### CODE REVIEW

File: cypress/e2e/identify.cy.js [hunk 1]
```diff
@@ -12,4 +12,56 @@ describe('identify()', () => {
...
+                'an-identified-event',
+                'another-identified-event',
+                'an-anonymous-event',
+                '$identify',
+                'an-identified-event',
+            ])
+
+            expect(new Set(events.map((e) => e.properties['$device_id'])).size).to.eql(1)
+
+            // the first two events share a distinct id
```
Comment (pauldambra) [inline cypress/e2e/identify.cy.js:45]: this is a pretty verbose test but I want to prove the chain of behaviour not just the steps in isolation

File: src/__tests__/posthog-core.identify.js [hunk 2]
```diff
@@ -0,0 +1,285 @@
+import { PostHog } from '../posthog-core'
+import { PostHogPersistence } from '../posthog-persistence'
+
+jest.mock('../gdpr-utils', () => ({
+    ...jest.requireActual('../gdpr-utils'),
+    addOptOutCheck: (fn) => fn,
+}))
+jest.mock('../decide')
+
+given('lib', () => Object.assign(new PostHog(), given.overrides))
```
Comment (pauldambra) [inline src/__tests__/posthog-core.identify.js:10]: most of these tests are copied almost as-is from posthog-core.js test file which was mahoosive

**Comment by pauldambra** [comment]
@mariusandra interesting coincidental example of why we don't want to reset device id by default https://posthog.slack.com/archives/C03P7NL6RMW/p1675452786802429

**Review by mariusandra — APPROVED** [review]
I think this makes sense

**Comment by pauldambra** [comment]
Local testing showing merge of anonymous events

Screenshot 2023-02-06 at 18 02 07
Screenshot 2023-02-06 at 18 01 08

**Comment by pauldambra** [comment]
And with back to back identifies not merging

Screenshot 2023-02-06 at 18 04 57

**Pull request merged**